File: Dynamic Programming/*Dungeon Game/solution_btm_up.ts
function calculateMinimumHP(dungeon: number[][]): number {
    let m = dungeon.length
    let n = dungeon[0].length
    let dp = Array(m).fill(0).map(() => Array(n))
    dp[m - 1][n - 1] = Math.max(0, -dungeon[m - 1][n - 1])
    for(let i = m - 2; i >= 0; i--) {
        dp[i][n - 1] = Math.max(0, dp[i+1][n-1] - dungeon[i][n-1])
    }
    for(let i = n - 2; i >= 0; i--) {
        dp[m - 1][i] = Math.max(0, dp[m-1][i + 1] - dungeon[m-1][i])
    }
    for(let i = m - 2; i >= 0; i--) {
        for(let j = n - 2; j >= 0; j--) {
            dp[i][j] = Math.max(0, Math.min(dp[i+1][j], dp[i][j+1]) - dungeon[i][j])
        }
    }
    return dp[0][0] + 1
    
};

File: Dynamic Programming/Minimum Number of Taps to Open to Water a Garden/solution_btm_up.py
class Solution:
    def minTaps(self, n: int, ranges: List[int]) -> int:
        # Needs to cover the whole range
        # dp[i] = number of taps to cover all the garden from i backward
        dp = [0] + [math.inf for i in range(n)]
        for i, r in enumerate(ranges):
            left, right = max(0, i - r), min(n, i + r)
            for j in range(left, right + 1):
                dp[j] = min(dp[j], dp[left] + 1) # dp[left] is minTaps till before left
        
        if dp[-1] == math.inf:
            return -1
        return dp[-1]
